spec/fastsync/version2/fsv2_scheduler.tla

```diff
@@ -0,0 +1,532 @@
+    \/ OnGlobalTimeoutTicker
+    \/ OnAddPeerEv
+    \/ OnTrySchedule
+    \/ OnStatusResponseEv
+    \/ OnBlockResponseEv
+    \/ OnRemovePeerEv
+    \/ OnPcBlockProcessed
+    \/ OnPcBlockVerificationFailure
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The system is the composition of the environment and the schedule                             *)
+(* ----------------------------------------------------------------------------------------------*)
+Init == turn = "environment" /\ InitEnv /\ InitSc
+
+FlipTurn == 
+  turn' = (
+    IF turn = "scheduler" THEN 
+      "environment" 
+    ELSE
+      "scheduler"
+  ) 
+
+Next == \* scheduler and environment alternate their steps (synchronous composition introduces more states)
+  /\ FlipTurn
+  /\ IF turn = "scheduler" THEN 
+       /\ NextSc
+       /\ inEvent' = noEvent
+       /\ UNCHANGED envRunning
+     ELSE
+       /\ NextEnv
+       /\ outEvent' = noEvent
+       /\ UNCHANGED <<scRunning, scheduler>>  
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Expected properties                                                                           *)
+(* ----------------------------------------------------------------------------------------------*)
+
+TypeOK ==
+    /\ turn \in {"scheduler", "environment"}
+    /\ inEvent \in InEvents
+    /\ envRunning \in BOOLEAN
+    /\ outEvent \in OutEvents
+    /\ scheduler \in [
+         initHeight: Heights \cup {ultimateHeight + 1},
+         height: Heights \cup {ultimateHeight + 1},
+         peers: SUBSET PeerIDs,
+         peerHeights: [PeerIDs -> Heights \cup {None}],
+         peerStates: [PeerIDs -> PeerStates],
+         blocks: SUBSET Heights,
+         blockStates: [Heights -> BlockStates],
+         pendingBlocks: [Heights -> PeerIDs \cup {None}],
+         receivedBlocks: [Heights -> PeerIDs \cup {None}]
+       ]
+       
+TerminationWhenNoAdvance ==
+  (WF_turn(FlipTurn) /\ <> (inEvent.type = "tNoAdvanceExp"))
+    => <> ~scRunning
+
+NoFailuresAndTimeouts ==
+    \* no peer removal
+    /\ inEvent.type /= "peerRemoveEv"
+    \* no invalid blocks
+    /\ inEvent.type /= "bcNoBlockResponse"
+    /\ inEvent.type /= "pcBlockVerificationFailure"
+NoGlobalTimeout == inEvent.type /= "tNoAdvanceExp"
+
+FiniteResponse ==
+    <>[] (inEvent.type
+         \in {"bcAddNewPeer", "bcStatusResponse", "bcBlockResponse", "pcBlockProcessed", "pcBlockVerificationFailure", "rTrySchedule"})
+
+\* all blocks from initHeight up to max peer height have been processed
+AllRequiredBlocksProcessed ==
+  LET maxH == maxHeightScheduler(scheduler) IN
+  LET unprocessedBlocks == {\A h \in scheduler.initHeight.. maxH: scheduler.blockStates[h] /= "blockStateProcessed"} IN
+  scheduler.height >= maxH /\ Cardinality(unprocessedBlocks) = 0
+
+GoodTermination ==
+  (WF_turn(FlipTurn) /\ []NoGlobalTimeout /\ []NoFailuresAndTimeouts /\ FiniteResponse)
```

Comment:
Why is the FiniteResponse needed here and what does it add?

Reply:
To check termination in the "good" scenarios where we have stable peers that send good blocks.

Reply:
But then why does it include the VerificationFailure event?